text/0000-RouteInfo-Metadata.md

```diff
@@ -0,0 +1,212 @@
+- Start Date: 2018-11-02
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# RouteInfo MetaData
+
+## Summary
+
+The RFC introduces the ability to associate application specific metadata with it's corresponding `RouteInfo` object. This also adds a `metadata` field to the `RouteInfo` which will be the return value of `buildRouteInfoMetadata` for it's corresponding `Route`.
+
+```js
+// app/route/profile.js
+import Route from '@ember/routing/route';
+import { inject } from '@ember/service';
+export default Route.extend({
+  user: inject('user'),
+  buildRouteInfoMetadata() {
+    return {
+      trackingKey: 'page_profile',
+      user: {
+        id: this.user.id,
+        type: this.user.type
+      }
+    }
+  }
+  // ...
+});
+```
+
+```js
+// app/services/analytics.js
+import Service, { inject } from '@ember/service';
+
+export default Service.extend({
+  router: inject('router'),
+  init() {
+    this._super(...arguments);
+    this.router.on('routeDidUpdate', (transition) => {
+      let { to, from } = transition;
+      let fromMeta = from.metadata;
+      let toMeta = to.metadata;
+      ga.sendEvent('pageView', {
+        from: fromMeta,
+        to: toMeta,
+        timestamp: Date.now(),
+      })
+    })
+  },
+  // ...
+});
+```
+
+## Motivation
+
+While the `RouteInfo` object is sufficient in providing developers metadata about the `Route` itself, it is not sufficient in layering on application specific metadata about the `Route`. This metadata could be anything from just a more domain specific name for a `Route` e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.
+
+## Detailed design
+
+### `buildRouteInfoMetadata`
+
+This optional hook is intended to be used as a way of letting the routing system know about any metadata associated with the route. It is guranteed to be called once before `routeWillChange` so that it can populate the `RouteInfo`s.
+
+#### `Route` Interface Extension
+
+```ts
+interface Route {
+  // ... existing public API
+  buildRouteInfoMetadata(): unknown
+}
+```
+
+### `RouteInfo.metadata`
+
+The `metadata` optional field on the `RouteInfo` will be populated with the return value of `buildRouteInfoMetadata`. If there is no metadata associated with the `Route`, the `metadata` field will be `null`.
+
+```ts
+interface RouteInfo {
+  // ... existing public API
+  metadata: Maybe<unknown>;
+}
+```
+
+This field will also be added to `RouteInfoWithAttributes` as it is just a super-set of `RouteInfo`.
+
+
+## How we teach this
+
+We feel that this a low-level primitive that will allow existing tracking addons to encapsulate. That being said the concept here is pretty simple: What gets returned from `buildRouteInfoMetadata` becomes the value of `RouteInfo.metadata` for that `Route`.
```

Comment:
I think we need to be prescriptive about appropriate dependencies for the `buildRouteInfoMetadata` hook, since this blocks routing.